### ISSUE TITLE
Hierarchical Parameter Scripting

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#13746]
## Summary of the new feature/enhancement

To complement **Feature Idea: Revisiting and extending the ParameterSet Experience** #12818
I would like to raise two dissenting ideas:

### 1. Reverse the set/parameter containment
Personally, I think that the current parameter notation is inside out
Currently, **parameters contain parameters sets**  (each Parameter holds multiple attributes with a `ParametersSetName`)
but it should be visa versa: **a parameter [*set*](https://en.wikipedia.org/wiki/Set_theory) should contain parameters.**
By turning this around, you might also support (recursive) parameter [subsets](https://en.wikipedia.org/wiki/Subset).
With all respect, adding additional configuration items (like groups and [ParameterMutexName](https://github.com/PowerShell/PowerShell/issues/5175)) might help to reduce the redundancy but eventually adds up to the complicity when combining them with the existing configuration items.
I even believe that *naming* parameters sets is inessential if using subsets.

### 2. Use *PowerShell* to build parameter sets.
Developers creating cmdlets are already quiet known with PowerShell and need to learn a complete different parameter syntax  to get them straight for their functions.
Apart from the fact that you might reduce the redundancy and complexity by using variables and subsets, there is also another advantage in moving to PowerShell scripting: I guess that changes to the current parameter notification (like e.g. groups) will require an update of the PowerShell engine therefor engineers who like to supply some backwards compatibility (with e.g. Windows PowerShell) for their cmdlets/functions might be reluctant to use any new parameter features provided through the current parameter notation. A PowerShell framework using the `DynamicParam` section, might only require installing a module to supply backwards compatibility in customer scripts.

## Proposed technical implementation details
Let's assume a `Draw-Shape`  cmdlet that requires the following parameter (set) constrains:

* There are 3 possible shapes:
  * **Circle** (which requires one `Radius` parameter)
  * **Rectangle** (which requires a `Width` and a `Height` parameter)
  * **Triangle** (which requires 3 - `Side1`, `Side2` and `Side3` - parameters)
* Each shape requires either of the (*mutually exclusive*) parameters below:
  * `Transparency`
  * `Opacity`
* And a optional color (default black) by either using
  *  a **RGB** scheme
      * `Red`
      * `Green`
      * `Blue`
  * or a **HSL** scheme
      * `Hue`
      * `Sat`
      * `Lum`

Using the current parameter notation, this requires:
<details>
  <summary>a lot of parameter definitions...</summary>

``` PowerShell
function Draw-Shape {
    [CmdletBinding()] param(
        [Parameter(Position = 0, ParameterSetName = 'CircleTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'CircleTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'CircleOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'CircleOpacityHSL', Mandatory = $True)]
        [int]$Radius,
        
        [Parameter(Position = 0, ParameterSetName = 'SquareTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'SquareTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'SquareOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'SquareOpacityHSL', Mandatory = $True)]
        [int]$Width,

        [Parameter(Position = 1, ParameterSetName = 'SquareTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'SquareTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'SquareOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'SquareOpacityHSL', Mandatory = $True)]
        [int]$Height,

        [Parameter(Position = 0, ParameterSetName = 'TriangleTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'TriangleTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'TriangleOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 0, ParameterSetName = 'TriangleOpacityHSL', Mandatory = $True)]
        [int]$Side1,

        [Parameter(Position = 1, ParameterSetName = 'TriangleTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'TriangleTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'TriangleOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 1, ParameterSetName = 'TriangleOpacityHSL', Mandatory = $True)]
        [int]$Side2,

        [Parameter(Position = 2, ParameterSetName = 'TriangleTransparencyRGB', Mandatory = $True)]
        [Parameter(Position = 2, ParameterSetName = 'TriangleTransparencyHSL', Mandatory = $True)]
        [Parameter(Position = 2, ParameterSetName = 'TriangleOpacityRGB', Mandatory = $True)]
        [Parameter(Position = 2, ParameterSetName = 'TriangleOpacityHSL', Mandatory = $True)]
        [int]$Side3,
        
        [Parameter(ParameterSetName = 'CircleTransparencyRGB',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'SquareTransparencyRGB',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'TriangleTransparencyRGB', Mandatory = $True)]
        [Parameter(ParameterSetName = 'CircleTransparencyHSL',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'SquareTransparencyHSL',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'TriangleTransparencyHSL', Mandatory = $True)]
        [int]$Transparency,

        [Parameter(ParameterSetName = 'CircleOpacityRGB',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'SquareOpacityRGB',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'TriangleOpacityRGB', Mandatory = $True)]
        [Parameter(ParameterSetName = 'CircleOpacityHSL',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'SquareOpacityHSL',   Mandatory = $True)]
        [Parameter(ParameterSetName = 'TriangleOpacityHSL', Mandatory = $True)]
        [int]$Opacity,

        [Parameter(ParameterSetName = 'CircleTransparencyRGB')]
        [Parameter(ParameterSetName = 'SquareTransparencyRGB')]
        [Parameter(ParameterSetName = 'TriangleTransparencyRGB')]
        [Parameter(ParameterSetName = 'CircleOpacityRGB')]
        [Parameter(ParameterSetName = 'SquareOpacityRGB')]
        [Parameter(ParameterSetName = 'TriangleOpacityRGB')]
        [int]$Red,

        [Parameter(ParameterSetName = 'CircleTransparencyRGB')]
        [Parameter(ParameterSetName = 'SquareTransparencyRGB')]
        [Parameter(ParameterSetName = 'TriangleTransparencyRGB')]
        [Parameter(ParameterSetName = 'CircleOpacityRGB')]
        [Parameter(ParameterSetName = 'SquareOpacityRGB')]
        [Parameter(ParameterSetName = 'TriangleOpacityRGB')]
        [int]$Green,

        [Parameter(ParameterSetName = 'CircleTransparencyRGB')]
        [Parameter(ParameterSetName = 'SquareTransparencyRGB')]
        [Parameter(ParameterSetName = 'TriangleTransparencyRGB')]
        [Parameter(ParameterSetName = 'CircleOpacityRGB')]
        [Parameter(ParameterSetName = 'SquareOpacityRGB')]
        [Parameter(ParameterSetName = 'TriangleOpacityRGB')]
        [int]$Blue,

        [Parameter(ParameterSetName = 'CircleTransparencyHSL')]
        [Parameter(ParameterSetName = 'SquareTransparencyHSL')]
        [Parameter(ParameterSetName = 'TriangleTransparencyHSL')]
        [Parameter(ParameterSetName = 'CircleOpacityHSL')]
        [Parameter(ParameterSetName = 'SquareOpacityHSL')]
        [Parameter(ParameterSetName = 'TriangleOpacityHSL')]
        [int]$Hue,

        [Parameter(ParameterSetName = 'CircleTransparencyHSL')]
        [Parameter(ParameterSetName = 'SquareTransparencyHSL')]
        [Parameter(ParameterSetName = 'TriangleTransparencyHSL')]
        [Parameter(ParameterSetName = 'CircleOpacityHSL')]
        [Parameter(ParameterSetName = 'SquareOpacityHSL')]
        [Parameter(ParameterSetName = 'TriangleOpacityHSL')]
        [int]$Sat,

        [Parameter(ParameterSetName = 'CircleTransparencyHSL')]
        [Parameter(ParameterSetName = 'SquareTransparencyHSL')]
        [Parameter(ParameterSetName = 'TriangleTransparencyHSL')]
        [Parameter(ParameterSetName = 'CircleOpacityHSL')]
        [Parameter(ParameterSetName = 'SquareOpacityHSL')]
        [Parameter(ParameterSetName = 'TriangleOpacityHSL')]
        [int]$Lum
    )
}
```

</details>

Using this prototype [`ParameterFunctions`](https://github.com/iRon7/Hierarchical-Parameter-Scripting) module, it requires the following knowledge. Sorry, there is no readme yet, but you can use the `-?` option to get the syntax and the description below:

There are 2 functions included in the module to create parameters and parameter-sets:
1. `New-Parameter` (alias `New-Param`, or `NPM`)
This function has 5 main parameters:
  * `[int]$Position`
  * `[type]$Type`
  * `[string]$Name`
  * `[object]$Default`
  * `[switch]$Mandatory`

Besides has a few parameters like `HelpMessage` to set certain parameter attributes.
It has **no `ParameterSetName`** option!

2. `New-ParameterSet` (alias `New-ParamSet`, or `NPS`)
Which has two parameters:
  * `[RuntimeDefinedParameterDictionary[]]$ParameterSets`
  * `[String]$SetName` *(Optional)*

There can only be **one** root (default) parameter set (or a single parameter which is quiet pointless).
Each set might contain multiple (optional or mandatory) parameters or multiple (recursive) parameter sets.

## Syntax
The basic [syntax notation](https://docs.informatica.com/data-integration/powercenter/10-1/command-reference/using-the-command-line-programs/syntax-notation.html):

### `[x]`
Optional parameter. The command runs whether or not you enter optional parameters.

```PowerShell
function FMO {
    [CmdletBinding()] param()
    DynamicParam {
        New-Param 0 ([string]) x
    }
}
```

### `[x|y]`
Select between optional parameters.

```PowerShell
function FMO {
    [CmdletBinding()] param()
    DynamicParam {
        New-ParamSet @(
            New-Param 0 ([string]) x
            New-Param 1 ([string]) y
        )
    }
}
```

### `<x>`
Required option. If you omit a required option, the command line program returns an error message.

```PowerShell
function Test {
    [CmdletBinding()] param()
    DynamicParam {
        New-Param 0 ([string]) x -Mandatory
    }
}
```

### `<x|y>`
Select between required options. For the command to run, you must select from the listed options. If you omit a required option, the command line program returns an error message.

```PowerShell
function Test {
    [CmdletBinding()] param()
    DynamicParam {
        New-ParamSet @(
            New-ParamSet @(
                New-Param ([string]) x -Mandatory
            )
            New-ParamSet @(
                New-Param ([string]) y -Mandatory
            )
        )
    }
}
```

Further steps to come to the final result below you can find in the [`Parameters.Test.ps1`](https://github.com/iRon7/Hierarchical-Parameter-Scripting/blob/main/Parameter.Tests.ps1):

```PowerShell
function Draw-Shape {
    [CmdletBinding()] param()
    DynamicParam {

        # Define the optional  RGB and HSL subsets
        $Color = @(
            New-ParamSet @(
                New-Param ([int]) Red
                New-Param ([int]) Green
                New-Param ([int]) Blue
            )
            New-ParamSet @(
                New-Param ([int]) Hue
                New-Param ([int]) Sat
                New-Param ([int]) Lum
            )
        )

        # Define the optional mutually exclusive Transparency/Opacity subsets
        $Tint = @(
            New-ParamSet @(
                New-Param ([int]) Transparency -Mandatory
                $Color
            )
            New-ParamSet @(
                New-Param ([int]) Opacity -Mandatory
                $Color
            )
        )

        # Create a (single) main set
        New-ParamSet @(
            New-ParamSet @(
                New-Param 0 ([int]) Radius -Mandatory
                $Tint
            )
            New-ParamSet @(
                New-Param 0 ([int]) Width  -Mandatory
                New-Param 1 ([int]) Height -Mandatory
                $Tint
            )
            New-ParamSet @(
                New-Param 0 ([int]) Side1 -Mandatory
                New-Param 1 ([int]) Side2 -Mandatory
                New-Param 2 ([int]) Side3 -Mandatory
                $Tint
            )
        )
    }
}
```

### Draw-Shape -?
```PowerShell
NAME
    Draw-Shape

SYNTAX
    Draw-Shape [-Radius] <int> -Transparency <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Radius] <int> -Opacity <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Width] <int> [-Height] <int> -Transparency <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Width] <int> [-Height] <int> -Opacity <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Side1] <int> [-Side2] <int> [-Side3] <int> -Transparency <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Side1] <int> [-Side2] <int> [-Side3] <int> -Opacity <int> [-Red <int>] [-Green <int>] [-Blue <int>] [<CommonParameters>]

    Draw-Shape [-Radius] <int> -Transparency <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]

    Draw-Shape [-Radius] <int> -Opacity <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]

    Draw-Shape [-Width] <int> [-Height] <int> -Transparency <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]

    Draw-Shape [-Width] <int> [-Height] <int> -Opacity <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]

    Draw-Shape [-Side1] <int> [-Side2] <int> [-Side3] <int> -Transparency <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]

    Draw-Shape [-Side1] <int> [-Side2] <int> [-Side3] <int> -Opacity <int> [-Hue <int>] [-Sat <int>] [-Lum <int>] [<CommonParameters>]
```

## Caveats
* It concerns a prototype and there might be quiet some bugs a limitation but I think it might eventually replace the current parameter notation
* Consider a (default) parameter [Provider](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_providers?view=powershell-7) (I am lacking the C# knowledge to set this up 😌)
* It might be slower than the current parameter notation (consider a `-Static` option which caches the `RuntimeDefinedParameterDictionary`)
* Default value is not work (yet)
* Dynamic parameters are not automatically put in variables, workaround:
```PowerShell
foreach ($Name in $PSBoundParameters.get_keys()) {
    if (!(Test-Path variable:$Name)) { Set-Variable -Name $Name -Value $PSBoundParameters[$Name] }
}
```

**Comment by iSazonov** [comment]
In general I like the idea to enhance the parameter defining - 

What about additional level? I mean  an optional parameter could require some mandatory/optional subparameters.
And what about C# notation?

**Comment by vexx32** [comment]
The main reason I'm unsure about this style of declaring params is that you have declare the same parameter multiple times if it happens to be in multiple sets. This creates a greater possibility of mistakes and more confusing UX if a parameter by the same name appears to behave multiple different ways depending on the other parameters specified with it.

I would love to see a rework of parameter declaration, but I don't think this is the right direction to go. 🤔

**Comment by ThomasNieto** [comment]
Looking at the command syntax it looks like this cmdlet is trying to do too much. A good rule of thumb that I've seen with PowerShell parameter set best practices is that if you're getting this many, maybe its time to create dedicated cmdlets. This many parameter sets is also confusing to the user as the syntax diagram is too complex.

For your example the cmdlet could be broken down into multiple cmdlets:

`Draw-Shape` gets broken down into three cmdlets:

* `Draw-Circle -Radius <int>`
* `Draw-Rectangle -Height <int> -Width <int>`
* `Draw-Triangle -Side1 <int> -Side2 <int> -Side3 <int>`

Dedicated color cmdlets can be created to define either of the two color schemes. Then a parameter set can be made for the `-RGBColor` or `-HSLColor` or to be a common parameter. The same can be done with `-Transparency` and `-Opacity`.

* `New-RGBColor -Red <int> -Blue <int> -Red <int>`
* `New-HSLColor -Hue <int> -Sat <int> -Lum <int>`

This can take the 12 parameter sets down to just a couple by going this route. There was also an old [RFC](https://github.com/PowerShell/PowerShell-RFC/blob/master/X-Withdrawn/RFC0001-Mutual-Exclusion-Parameters-and-Properties.md) that defined mutually exclusive parameters without needing a new parameter set.

**Comment by iSazonov** [comment]
I added more info in #12818.

**Comment by iRon7** [comment]
@Vexx32 and @ThomasNieto, the 'Draw-Shape' is just an extensive example which two mutually exclusive parameter sets almost at the root of the main set.

> The main reason I'm unsure about this style of declaring params is that you have declare the same parameter multiple times if it happens to be in multiple sets. This creates a greater possibility of mistakes and more confusing UX if a parameter by the same name appears to behave multiple different ways depending on the other parameters specified with it.

Yes, there are about **22 (sub)sets in the `Draw-Shape` example with a total of 58 parameters** (where the current parameters notation has **14 parameters with 72 parameters-sets**). As @iSazonov pointed out in #12818, my general objective is to use **a syntax diagram which is similar to the (expanded) syntax displayed by the help**. In the existing solution, you might limit some of the parameter-sets by using configuration items like `DefaultParameterSetName` but it is difficult to further reduce the redundancy of this rather flat design. This is mainly because the attributes that *usually* relate to the parameter (like `position` and `mandatory`) are contaminated with a `ParameterSetName` property that actually belongs to parameter-set. In the purposed design it is easy to script any redundancy using variables or loops:

Taking the (current) existing [`Where-Object`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/where-object?view=powershell-7) cmdlet for another example:

```PowerShell
function WhereObject {
    [CmdletBinding()] param()
    DynamicParam {
        $MandatoryOperators = @(
            'CEQ', 'NE', 'CNE', 'GT', 'CGT', 'LT', 'CLT', 'GE', 'CGE', 'LE', 'CLE',
            'Like', 'CLike', 'NotLike', 'CNotLike', 'Match', 'CMatch', 'NotMatch', 'CNotMatch',
            'Contains', 'CContains', 'NotContains', 'CNotContains',
            'In', 'CIn', 'NotIn', 'CNotIn', 'Is', 'IsNot'
        )
        New-ParamSet @(
            New-ParamSet @(
                New-ParamSet @(
                    New-Param 0 ([scriptblock]) FilterScript -Mandatory
                    New-Param ([psobject]) InputObject
                )
            )
            New-ParamSet @(
                New-Param 0 ([string]) Property -Mandatory
                New-ParamSet @(
                    New-Param 1 ([object]) Value
                    New-Param ([switch]) EQ
                    New-Param ([psobject]) InputObject
                )
                foreach ($Operator in $MandatoryOperators) {
                    New-ParamSet @(
                        New-Param 1 ([object]) Value
                        New-Param ([switch]) $Operator -Mandatory
                        New-Param ([psobject]) InputObject
                    )
                }
            )
            New-ParamSet @(
                New-Param 0 ([string]) Property -Mandatory
                New-ParamSet @(
                    New-Param ([switch]) Not -Mandatory
                    New-Param ([psobject]) InputObject
                )
            )
        )
    }
}
```

**Comment by vexx32** [comment]
Yeah, you can do that with `dynamicparam` but you wouldn't likely be able to do anything remotely similar if we had builtin syntax for this (good luck parsing that; it'd be insanely complicated to try to make that work in anything like a `param` block).

Also... just in general, dynamicparam tends to cause more problems that it solves, so IMO any design _based_ on it is generally best avoided (even when prototyping other designs). YMMV I suppose, but I don't think this is an approachable angle for users.

**Comment by iRon7** [comment]
> What about additional level? I mean an optional parameter could require some mandatory/optional subparameters.

At first sight, you might think of:
```PowerShell
function Test {
    [CmdletBinding()] param()
    DynamicParam {
        New-ParamSet @(
            New-Param ([string]) Opt1
            New-ParamSet @(
                New-Param ([string]) Man1 -Mandatory
                New-Param ([string]) Opt2
            )
        )
    }
}
```
But that actually means: **`[ [Opt1] | [Man1 | [Opt2] ]`** and consequently results in:
```
SYNTAX
    test -Man1 <string> [-Opt2 <string>] [<CommonParameters>]
    test [-Opt1 <string>] [<CommonParameters>]
```

I am still trying define some rules here...
Concerning design of the prototype module: I am putting each **mandatory** parameter in a separate set together with all the parameters in each subset. This means that the mandatory parameters in a subset will only be effective if there is a mandatory parameter in the parent. ...All the way up to the root (I was thinking of throwing an error if there is a mandatory parameter in the subset and not in the parent but didn't implement this yet).

This means that the 'optional" parameter in the parent actually needs to be mandatory (which make sense because if you use one of the parameters in the subset, the parameter in the parent set isn't anymore optional).
In other words, the correct syntax for this is **`[ Opt1 Man1 | Opt1 [Opt2] ]`**:
```PowerShell
function Test {
    [CmdletBinding()] param()
    DynamicParam {
        New-ParamSet @(
            New-Param ([string]) Opt1 -Mandatory
            New-ParamSet @(New-Param ([string]) Man1 -Mandatory)
            New-ParamSet @(New-Param ([string]) Opt2)
        )
    }
}
```
Which also shows from the help syntax:
```
SYNTAX
    test -Man1 <string> -Opt1 <string> [<CommonParameters>]
    test -Opt1 <string> [-Opt2 <string>] [<CommonParameters>]
```

> And what about C# notation?

Unfortunately, I do not have enough C# experience to comment on this.

**Comment by iRon7** [comment]
@vexx32 

> you can do that with `dynamicparam` but you wouldn't likely be able to do anything remotely similar if we had builtin syntax for this.

Agree, it is not in the intention to align the current `Param()` notation with the purpose. Besides, as already noted, that would require an engine update where the `dynamicparam` purpose might only require a module which could even supply some downward compatibility with older (Windows) PowerShell versions.

> `dynamicparam` tends to cause more problems that it solves

I think, that is mainly because there is currently no proper framework (or provider) and guideline to support the developers in using this.

**Comment by iSazonov** [comment]
> I was thinking of throwing an error if there is a mandatory parameter in the subset and not in the parent but didn't implement this yet

I'd expect symmetric behavior for all levels.

> And what about C# notation?
> Unfortunately, I do not have enough C# experience to comment on this.

Script and binary developers should get the same capabilities otherwise your work will be only partial research.

**Comment by iRon7** [comment]
> I'd expect symmetric behavior for all levels.

That depends on the actual implementation, it this purpose I trying to follow the syntax diagram*, where subsets (`[...]`) are not just extensions of the parent set but **exclusive** optional sets or in the case all subsets contain at least one mandatory parameter, they are considered ***mutually* exclusive** sets. And in case there is just one subset the subset is mandatory if it contains at least one mandatory parameter and optional if there are no mandatory parameters in the subset.

* *Which the exception where it concerns the definition of a single parameter:
    * An optional parameter `[MyParam]` is created like: `New-Param MyParam`
    * A mandatory parameter `MyParam` is created like: `New-Param MyParam -Mandatory`

**Comment by iRon7** [comment]
I have updated the [prototype module](https://github.com/iRon7/Hierarchical-Parameter-Scripting) to better support optional parameters (@iSazonov "_an optional parameter could require some mandatory/optional subparameters_" comment) which also results in a more symmetric behavior (@vex32 "_I'd expect symmetric behavior for all levels_" comment).

## (Sub)set design
Below is the description how the `ParameterSetName` of each parameter is set with respect to their (sub)set containers.

* If there is no `SetName` supplied, the `ParameterSetName`  is initially set to `'0'` (a string containing a zero).
* The `SetName` is added to:
    * the `ParameterSetName` attribute of each optional parameter
    * the `ParameterSetName` attribute of each mandator parameter _unless the set also contains subsets_
* In case of a subset, the `ParameterSetName` of the each attribute of each containing parameter is prefixed `SetName` (or `'0'`, if no setname is supplied) separated by a dot `'.'`.
    * Any `ParameterSetName` parameter attribute in a subset that is already prefixed with a `'0'` (e.g. `'0.1'`) will get a new unique index prefix (e.g. `'1.1'`). In de case of a _named_ subset, the developer is responsible to supply unique names for each subset.

Note that with each subset iteration, the whole parameter set (including the containing parameter attributes) is cloned. This is to support parameter sets supplied in variable (which eventually require an unique `ParameterSetName`).

## (Sub)set usage:
The basic idea of this purpose is to follow the standard parameter syntax diagram similar to the Help (`-?`) syntax output..

* There is one exception where it concerns the definition of a _single_ parameter:
    * An optional parameter `[MyParam]`, is created like: `New-Param MyParam`
    * A mandatory parameter `MyParam`, is created like: `New-Param MyParam -Mandatory`

* Every parameter subset is **exclusive**
  This means that a user can select a specific subset (and supply the contained mandatory parameters) but _can't_ choose parameters out of different subsets
    * If the parent set and _all_ subsets contain at least one mandatory parameter, the parameter subset is **mutually exclusive**
    This means that a user _must_ choose one specific subset
    * If the parent set or one of the subsets contain _only_ optional parameters the parameter subset is an **exclusive choose**
    This means that the user _might_ to choose one subset (or none of the subsets)

* The user is required to supply both mandatory parameters of the selected subset and the mandatory parameters of its parent set.
